### PR TITLE
chore(hadron-build): automate download center manifest update and upload

### PR DESCRIPTION
**Pull request by gribnoysup**
This patch update `hadron-build upload` command to

- Upload all target variations in one go (this is how we do this in general, so no reason to support partial uploads, it just complicates things)
- Update download center manifest file with a provided version by using the target variations and not just replacing the version number in the existing URL

TODO:

- [x] I'll add some unit tests for separate parts of the update logic (as for testing the full thing I don't think anything but doing a beta release will do here)

### CODE REVIEW

File: packages/compass/release/publish.js
Comment (gribnoysup) [inline packages/compass/release/publish.js:66]: `cli.error` will throw, so this `info` was unreachable

File: packages/hadron-build/lib/target.js
Comment (gribnoysup) [inline packages/hadron-build/lib/target.js:627]: This doesn't seem necessary based on [deb version policies](https://www.debian.org/doc/debian-policy/ch-controlfields.html#version) and was removed when uploading release assets to github anyway, so seemed like a good idea to remove this. Please tell me if you think otherwise

File: packages/hadron-build/lib/target.js
Comment (gribnoysup) [inline packages/hadron-build/lib/target.js:365]: We don't list all the assets in the download center, so this explicitly marks the ones that will be shown there

File: packages/hadron-build/lib/download-center.js
Comment (gribnoysup) [inline packages/hadron-build/lib/download-center.js:31]: Before we were skipping uploads for existing files, I removed this check because this just doesn't sound right to me: if we would need to re-run asset publishing (which is separate from making assets listed on downloads page) I think we would want the latest version to be available. But please tell me if this doesn't make sense to y'all

File: packages/hadron-build/lib/download-center.js
Comment (gribnoysup) [inline packages/hadron-build/lib/download-center.js:71]: So this manifest would actually get uplaoded on every release, but it's not the one that is used for the download page and I am not really sure what it is in general: it's pretty outdated, download links are not correct and half of the supported OS is missing, so I just removed this. If anyone knows what this is used for, please tell!

File: packages/compass/release/publish.js
Comment (mcasimir) [inline packages/compass/release/publish.js:41]: I would just remove `npm run release publish`, at this point everything from changelog to publishing the manifest is outdated, we don't really need it anymore.

Means we can ditch also the changelog code and many more

Comment (mcasimir) [reply packages/compass/release/publish.js:41]: I can do the cleanup in a follow-up PR if you want, we should also get rid of the checks that still require AWS keys locally

File: packages/hadron-build/lib/download-center.js
Comment (mcasimir) [reply packages/hadron-build/lib/download-center.js:71]: This thing was disabled at some point, not sure why, is safe to be removed

File: packages/hadron-build/lib/target.js
Comment (mcasimir) [reply packages/hadron-build/lib/target.js:627]: yeah, looks like we don't do that for mongosh

File: packages/hadron-build/lib/download-center.js
Comment (mcasimir) [inline packages/hadron-build/lib/download-center.js:44]: Where is this called? and with what manifest?

**Review by mcasimir — COMMENTED** [review]
Looks great so far, i can't find the part that actually builds and upload the manifest, how does it work?

File: packages/compass/release/publish.js
Comment (gribnoysup) [reply packages/compass/release/publish.js:41]: Oh, I actually thought that the changelog for the releases is still something that you need to manually apply in GitHub. is it not true anymore? How does the changelog message gets set now?

I can do a more through clean up in the same PR, thanks for highlighting! I just genuinely thought that this is still required 😅

**Comment by gribnoysup** [comment]
@mcasimir (and anyone else who wants to review this) github collapsing files strikes again! The updated upload script is in the collapsed [packages/hadron-build/commands/upload.js](https://github.com/mongodb-js/compass/pull/3299/files#diff-1f536797a749337fcbb568c4c9b37523466a7d5118cbfb21af737b40833ed70d), there is a bunch of changes there so might make sense to review the code itself and don't pay too much attention to the diff

File: packages/compass/release/publish.js
Comment (gribnoysup) [reply packages/compass/release/publish.js:41]: Ahhh, it's this GitHub releases thingie, now I remember 🤦

File: packages/hadron-build/commands/upload.js
Comment (mcasimir) [inline packages/hadron-build/commands/upload.js:214]: Not necessary for this PR, but I think we can improve things further.

It would be amazing to have a "sync" function with what's published on Github rather than "upload(version)".

That's a function that you would be able to call also to "unpublish" a release and wouldn't depend on what was uploaded before.

Comment (mcasimir) [reply packages/hadron-build/commands/upload.js:214]: Btw here we may want to check that the version we update is more recent than the previous one, i guess is not super common, but since it doesn't take the latest, if i unpublish and re-publish an old version in github that would replace the latest one right?

Comment (mcasimir) [reply packages/hadron-build/commands/upload.js:214]: Just occurred to me that we could do something different:

- as part of EVG upload, we upload a `manifest.json` to Github and the hadron-build job ends here
- on the github workflow each time a release is published or unpublished we get the manifest from the latest GA and latest beta on github, merge them and upload it to S3

Comment (gribnoysup) [reply packages/hadron-build/commands/upload.js:214]: Ah, I actually like this idea a lot and it wouldn't be that hard to refactor what I have here already to this, let's do it that way!

File: packages/hadron-build/commands/upload.js
Comment (gribnoysup) [inline packages/hadron-build/commands/upload.js:354]: Not sure if we should fail here instead or not

File: packages/hadron-build/commands/upload.js
Comment (gribnoysup) [reply packages/hadron-build/commands/upload.js]: Updated to upload manifest as part of gh release and download latest stable / beta to build updated manifest on manifest update command

File: packages/hadron-build/lib/target.js
Comment (addaleax) [inline packages/hadron-build/lib/target.js:804]: is this a typo?
```suggestion
    const match = version.match(/-([a-z]+)(\.\d+)?$/);
    if (match) {
      return match[1].toLowerCase();
```

File: packages/hadron-build/lib/target.js
Comment (addaleax) [inline packages/hadron-build/lib/target.js:785]: While this works just fine, any reason not to use `.flatMap()` in these places?

File: packages/hadron-build/lib/target.js
Comment (gribnoysup) [reply packages/hadron-build/lib/target.js:804]: I copy pasted it from old code without changes and I think someone was just really into unnecessary variable name shortening 😆 Will update!

File: packages/hadron-build/lib/target.js
Comment (gribnoysup) [reply packages/hadron-build/lib/target.js:785]: That's a great suggestion! I just need to build some muscle memory to start using it, that's the only reason I'm not using it here, will update 😄

File: packages/hadron-build/commands/upload.js
Comment (gribnoysup) [reply packages/hadron-build/commands/upload.js:113]: hadron-build is on the "old" eslint config 😞 I didn't want to make the diff more convoluted, but will do a follow up and update it to the new one (it also will allow us to eventually move the code here to typescript) and will clean up unused eslint rules

File: packages/hadron-build/commands/upload.js
Comment (gribnoysup) [reply packages/hadron-build/commands/upload.js:314]: Yep, good catch. I was playing with the idea of getting the diff from nodejs deepStrictEqual assertion error, but eh... it was kinda weird and so I decided to use a library (would be cool though if this string diff was something that is assigned to the assertion error separately from the message maybe 🤔)

File: packages/hadron-build/commands/upload.js
Comment (addaleax) [reply packages/hadron-build/commands/upload.js:113]: Ah no worries – if we’re already using a newer eslint config elsewhere, that also sounds good. I just felt like I’d run into this rule occasionally in the past and assumed that it would affect Compass more broadly, that’s all

File: .evergreen/functions.yml
Comment (gribnoysup) [reply .evergreen/functions.yml:285]: Oh, didn't know that it's a common was of doing this, makes sense, will update!

File: packages/hadron-build/lib/download-center.js
Comment (gribnoysup) [reply packages/hadron-build/lib/download-center.js:71]: Aha, good to know. Well, it's completely removed now

File: packages/hadron-build/commands/upload.js
Comment (gribnoysup) [reply packages/hadron-build/commands/upload.js]: Removed

File: packages/hadron-build/lib/target.js
Comment (gribnoysup) [reply packages/hadron-build/lib/target.js]: Updated, much cleaner now, ty!

**Comment by gribnoysup** [comment]
[Successful dry run publish](https://evergreen.mongodb.com/task_log_raw/10gen_compass_main_ubuntu_publish_publish_patch_8b6b620a0a36ba956a5a2e3d4ea73603e4730c92_62e7b6e22fbabe511ca0dbe6_22_08_01_11_20_11/0?type=T&text=true)

File: packages/hadron-build/lib/download-center.js
Comment (mcasimir) [inline packages/hadron-build/lib/download-center.js:42]: Is this still used? Can we clean it up or do we plan to do something with it?

File: packages/hadron-build/lib/target.js
Comment (mcasimir) [inline packages/hadron-build/lib/target.js:651]: how did this work ever? did it got renamed when we uploaded on the internal S3?

File: packages/hadron-build/commands/upload.js
Comment (mcasimir) [inline packages/hadron-build/commands/upload.js:339]: I really think we shouldn't do this and rather generate a fresh new one ignoring anything that was there before. Any particular reason why we want to rather modify the old one?

File: packages/hadron-build/lib/download-center.js
Comment (gribnoysup) [reply packages/hadron-build/lib/download-center.js:42]: Yes, this is used to diff manifests before update, print the diff for debugging purposes, warn if they are the same (which is an unexpected thing). See https://github.com/mongodb-js/compass/pull/3299#discussion_r933258183 and feel free to tell me whether it is okay to just warn in this case

File: packages/hadron-build/lib/target.js
Comment (gribnoysup) [reply packages/hadron-build/lib/target.js:651]: It did not for awhile now, but also these files are not shown in download center and probably nobody ever used them, so we never got any error reports about this

File: packages/hadron-build/commands/upload.js
Comment (mcasimir) [reply packages/hadron-build/commands/upload.js:346]: I like that, seems a legit case for example if we publish/unpublish old versions

File: packages/hadron-build/commands/upload.js
Comment (gribnoysup) [reply packages/hadron-build/commands/upload.js:339]: I don't really have a preference so will change. There is only a handful of metadata fields here so it doesn't seem that important if we use the old ones or just have them hardcoded in this file

File: packages/hadron-build/commands/upload.js
Comment (mcasimir) [reply packages/hadron-build/commands/upload.js:346]: Though this code confused me a tons, cause it throws in the "happy path", would it be ok to extract the check to a function that doesn't throw?

```
isDeepStrictEqual =  (obj1, obj2) => {
   try { deepStrictEqual(obj1, obj2); return true; }  catch { return false; }
}

if (isDeepStrictEqual(m1, m2)) {
   return;
}

// actually upload

```

File: packages/hadron-build/lib/download-center.js
Comment (mcasimir) [reply packages/hadron-build/lib/download-center.js:42]: I found it, it was in a collapsed file

File: packages/hadron-build/commands/upload.js
Comment (gribnoysup) [reply packages/hadron-build/commands/upload.js]: Updated in 577bb263683a7bccad62f5cda0656ffc39262f53

Comment (mcasimir) [reply packages/hadron-build/commands/upload.js]: Thanks! I like it a lot better to have the entire file here and avoid any mutation, like that we can control everything in this place and if we would ever need to change it we can do it with a PR

**Review by mcasimir — APPROVED** [review]

File: packages/hadron-build/commands/upload.js
Comment (gribnoysup) [reply packages/hadron-build/commands/upload.js]: Yep, this sounds like a good reason!

**Comment by gribnoysup** [comment]
Waiting for the new GA to be finished and will merge after it is done

**Pull request merged**
